test/common/upstream/logical_dns_cluster_test.cc

```diff
@@ -50,6 +65,101 @@ class LogicalDnsClusterTest : public testing::Test {
         }));
   }
 
+  void testBasicSetup(const std::string& config, ConfigType config_type = ConfigType::V2_YAML) {
+    expectResolve(Network::DnsLookupFamily::V4Only);
+    if (config_type == ConfigType::V1_JSON) {
+      setupFromV1Json(config);
+    } else {
+      setupFromV2Yaml(config);
+    }
+
+    EXPECT_CALL(membership_updated_, ready());
+    EXPECT_CALL(initialized_, ready());
+    EXPECT_CALL(*resolve_timer_, enableTimer(std::chrono::milliseconds(4000)));
+    dns_callback_(TestUtility::makeDnsResponse({"127.0.0.1", "127.0.0.2"}));
+
+    EXPECT_EQ(1UL, cluster_->prioritySet().hostSetsPerPriority()[0]->hosts().size());
+    EXPECT_EQ(1UL, cluster_->prioritySet().hostSetsPerPriority()[0]->healthyHosts().size());
+    EXPECT_EQ(0UL,
+              cluster_->prioritySet().hostSetsPerPriority()[0]->hostsPerLocality().get().size());
+    EXPECT_EQ(
+        0UL,
+        cluster_->prioritySet().hostSetsPerPriority()[0]->healthyHostsPerLocality().get().size());
+    EXPECT_EQ(cluster_->prioritySet().hostSetsPerPriority()[0]->hosts()[0],
+              cluster_->prioritySet().hostSetsPerPriority()[0]->healthyHosts()[0]);
+    HostSharedPtr logical_host = cluster_->prioritySet().hostSetsPerPriority()[0]->hosts()[0];
+
+    EXPECT_CALL(dispatcher_,
+                createClientConnection_(
+                    PointeesEq(Network::Utility::resolveUrl("tcp://127.0.0.1:443")), _, _, _))
+        .WillOnce(Return(new NiceMock<Network::MockClientConnection>()));
+    logical_host->createConnection(dispatcher_, nullptr);
+    logical_host->outlierDetector().putHttpResponseCode(200);
+
+    expectResolve(Network::DnsLookupFamily::V4Only);
+    resolve_timer_->callback_();
+
+    // Should not cause any changes.
+    EXPECT_CALL(*resolve_timer_, enableTimer(_));
+    dns_callback_(TestUtility::makeDnsResponse({"127.0.0.1", "127.0.0.2", "127.0.0.3"}));
+
+    EXPECT_EQ(logical_host, cluster_->prioritySet().hostSetsPerPriority()[0]->hosts()[0]);
+    EXPECT_CALL(dispatcher_,
+                createClientConnection_(
+                    PointeesEq(Network::Utility::resolveUrl("tcp://127.0.0.1:443")), _, _, _))
+        .WillOnce(Return(new NiceMock<Network::MockClientConnection>()));
+    Host::CreateConnectionData data = logical_host->createConnection(dispatcher_, nullptr);
+    EXPECT_FALSE(data.host_description_->canary());
+    EXPECT_EQ(&cluster_->prioritySet().hostSetsPerPriority()[0]->hosts()[0]->cluster(),
+              &data.host_description_->cluster());
+    EXPECT_EQ(&cluster_->prioritySet().hostSetsPerPriority()[0]->hosts()[0]->stats(),
+              &data.host_description_->stats());
+    EXPECT_EQ("127.0.0.1:443", data.host_description_->address()->asString());
+    EXPECT_EQ("", data.host_description_->locality().region());
```

Comment:
Do we have any tests for the actual locality behavior?

Reply:
Not yet. I'll try to provide it. For `STATIC` and `STRICT_DNS` it seems obvious to have multiple localities, however, I couldn't wrap my head around configuring `LOGICAL_DNS` with multiple localities (and priorities; how to actually group the resolved hosts into different localities since the configuration of it is attached to the given domain). If you have some hints, I would love to hear that. Thanks!

Reply:
I think I have an idea now (and the current state in the last commit is not right), will update soon.

Reply:
From the code, I can see that a cluster with `LOGICAL_DNS` discovery type has only one "logical host" holder.
Does it make sense to have multiple `locality_lb_endpoints`?
Then we need to make sure it is only one `lb_endpoint` per `locality_lb_endpoint`. Consequently, it needs multiple logical hosts.

The new configuration (for an embedded load assignment in CDS) reflecting the above is as follows:

```yaml
    type: LOGICAL_DNS

    ...

    load_assignment:
      endpoints:
      - priority: 0
        lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: primary.bar.com
                port_value: 11001

      - priority: 1
        lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: backup.bar.com
                port_value: 11003
```

Reply:
I haven't looked through the change yet, will do so now, but I think it's going to be pretty odd for LOGICAL_DNS to have any concept of locality. I'm not sure how it would be used. Priority does potentially make sense in the face of health checking, but TBH, I'm not sure it's worth the churn and regression risk vs. just having a config check that it's not supported for now until someone actually asks for it / has a good design.

Reply:
Nevermind, I think you implemented what I had in mind. This change is pretty complicated but looks very cool, so i will carve out some more dedicated time tomorrow to look through it. :)